### PR TITLE
[expo-branch] First 3rd party library unimodule wrapper!

### LINKED ISSUE #5132: Research and implement Branch as an optional module on Android
Recently we've removed Branch from Android (more details [in the blog post](https://blog.expo.io/changes-to-expo-branch-support-d002c4bc564e)) but we'd like to provide a way to include this module if someone wants to do a build with it, maybe as a `--include-branch` flag to `expo build:android` command. 
Pull request removing Branch code for reference: #4058

### PR DESCRIPTION
**Pull request by sjchmiela**
# Why

We would like to provide a way to users to optionally include `react-native-branch` in their standalone apps. Fixes https://github.com/expo/expo/issues/5132.

# How

I added support for React Native-specific unimodules to React Native adapter. Right now it works like this:

- unimodule `expo-branch` depends on `react-native`, so it can use its API. It's no longer a uni-platform module, more an optinmodule (I guess the name won't stick)
- we copy `react-native-branch` vendored module directly to the unimodule. It will compile, since we added necessary dependencies to the unimodule
- `BranchPackage < org.unimodules.Package` implements `ReactPackage` interface too, forwarding all the calls to the vendored `RNBranchPackage`.
- when the module registry provider creates a module registry from a list of `Packages`, it can filter out `ReactPackages` — and so it does, putting them into a local internal module (`ReactPackagesProvider`) which only purpose is to carry those to `ModuleRegistryAdapter` which is responsible for creating a list of React Native modules out of a module registry. The adapter grabs the `ReactPackagesProvider` and adds React Native modules to the list.

# Test Plan

I have confirmed that adding `BranchPackage` to a list of installed unimodules' packages creates `RNBranchModule` and exports to JS.

### CODE REVIEW

File: packages/expo-branch/unimodule.json [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{
+  "name": "expo-branch",
+  "platforms": ["android"],
+  "android": {
+    "packages": ["expo.modules.branch.BranchPackage"]
```
Comment (wkozyra95) [inline packages/expo-branch/unimodule.json]: it's not required to specify package here
https://github.com/unimodules/react-native-unimodules/blob/master/gradle.groovy#L61

**Review by wkozyra95 — APPROVED** [review]

File: packages/expo-branch/unimodule.json (hunk 1 above)
Comment (sjchmiela) [reply packages/expo-branch/unimodule.json]: Apart from `BranchPackage < org.unimodules.Package` we also have `vendored.RNBranchPackage < com.facebook.react.Package` in the same module, which introduces a conflict. Both removing the first one or modifying the second one would require manual changes when updating the vendored code.

https://github.com/unimodules/react-native-unimodules/blob/09234cff13d179c74cf347136c04a477621c9fc7/gradle.groovy#L66-L68

**Review by esamelson — COMMENTED** [review]
awesome, this is really exciting -- thanks for this! 😄 what's here looks good, just a couple of comments--
- would be nice to add some internal documentation around this, e.g. how to add a new module this way, how to keep the branch module up to date, anything extra that needs to be done in the release process, etc. i think our internal docs are on the sparser side and just want to make sure we're taking opportunities to improve them!
- could we use expo-branch in ios as well? i know we can't do optional modules yet on ios, but it would be nice not to have branch (and other third party modules) in two different places in the repo.

**Comment by sjchmiela** [comment]
As @esamelson proposed, I moved Branch code from iOS `Exponent/Versioned` to `expo-branch`. This allowed me to create a full-fledged `expo-branch` unimodule, which cleanly integrates with properly configured `AppDelegate`s (using `UMAppDelegateWrapper`). It also allowed me to remove Expo-specific changes from `RNBranch` in favor of subclassing it in `EXScopedBranch` belonging to `ExpoKit` (bare apps will get unchanged `RNBranch` wired up with `expo-branch/EXBranchManager` which removes any need for extra setup steps).

`expo-branch` code is easily updated using `expotools`:
```
et update-module -m react-native-branch -p ${platform} -c ${commit}
```

**Review by tsapeta — APPROVED** [review]

**Pull request merged**